### PR TITLE
Add TPM measurement support

### PR DESCRIPTION
**Pull request by mjg59**
Add support for measuring the various boot components

### CODE REVIEW

File: grub-core/boot/i386/pc/diskboot.S [hunk 1]
```diff
@@ -295,6 +312,29 @@ LOCAL(copy_buffer):
 /* END OF MAIN LOOP */
 
 LOCAL(bootit):
+#ifdef TPM
+	pusha
+	movw	$0xBB07, %ax		/* TCG_CompactHashLogExtendEvent */
+
+	movw	$0x0, %bx
```
Comment (vcaputo) [inline grub-core/boot/i386/pc/diskboot.S]: any reason you're not using xor to zero here as done above?

File: grub-core/kern/efi/tpm.c [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+      return grub_error (GRUB_ERR_OUT_OF_MEMORY,
+			 N_("cannot allocate TPM event buffer"));
+    }
+
+    event->pcrindex = pcr;
+    event->eventtype = 0x0d;
+    event->eventsize = grub_strlen(description) + 1;
+    grub_memcpy(event->event, description, event->eventsize);
+
+    algorithm = 0x00000004; /* SHA 1 */
```
Comment (vcaputo) [inline grub-core/kern/efi/tpm.c]: can we give these magic numbers symbolic names in the form of #defines or something? (the 0x0d above as well)

File: grub-core/kern/efi/tpm.c [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    tpm2 = grub_efi_open_protocol (tpm_handle, &tpm2_guid,
+				   GRUB_EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+
+    if (!grub_tpm2_present(tpm2)) {
+      return 0;
+    }
+
+    event2 = grub_zalloc(sizeof (Event) + grub_strlen(description) + 1);
+    if (!event2) {
+      grub_printf("No buffer\n");
```
Comment (vcaputo) [inline grub-core/kern/efi/tpm.c]: residual debugging?

File: grub-core/kern/i386/pc/tpm.c [hunk 4]
```diff
@@ -0,0 +1,132 @@
+#include <grub/err.h>
+#include <grub/i18n.h>
+#include <grub/mm.h>
+#include <grub/tpm.h>
+#include <grub/misc.h>
+#include <grub/i386/pc/int.h>
+
+int tpm_present (void);
```
Comment (vcaputo) [inline grub-core/kern/i386/pc/tpm.c]: inconsistent formatting with the rest of the function signatures, especially 2 lines down, why all the spaces?

File: grub-core/kern/i386/pc/tpm.c [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  regs.ecx = 0;
+  regs.edx = 0;
+  regs.es = (inaddr & 0xffff0000) >> 4;
+  regs.edi = inaddr & 0xffff;
+  regs.ds = outaddr >> 4;
+  regs.esi = outaddr & 0xf;
+
+  grub_bios_interrupt (0x1a, &regs);
+
+  if (regs.eax) {
```
Comment (vcaputo) [inline grub-core/kern/i386/pc/tpm.c]: unnecessary curly braces

File: grub-core/kern/i386/pc/tpm.c [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	EventOutgoing outgoing;
+	Event *event;
+	grub_uint32_t datalength;
+
+	if (!tpm_present())
+		return 0;
+
+	datalength = grub_strlen(description);
+	event = grub_zalloc(datalength + sizeof(Event));
+	if (!event) {
```
Comment (vcaputo) [inline grub-core/kern/i386/pc/tpm.c]: curly braces

File: grub-core/kern/i386/pc/tpm.c [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	regs.es = (((grub_addr_t) &incoming) & 0xffff0000) >> 4;
+	regs.edi = ((grub_addr_t) &incoming) & 0xffff;
+	regs.ds = (((grub_addr_t) &outgoing) & 0xffff0000) >> 4;
+	regs.esi = ((grub_addr_t) &outgoing) & 0xffff;
+
+	grub_bios_interrupt (0x1a, &regs);
+
+	grub_free(event);
+
+	if (regs.eax) {
```
Comment (vcaputo) [inline grub-core/kern/i386/pc/tpm.c]: curly braces

File: grub-core/loader/i386/efi/linux.c [hunk 8]
```diff
@@ -340,6 +319,9 @@ grub_cmd_linux (grub_command_t cmd __attribute__ ((unused)),
   if (file)
     grub_file_close (file);
 
+  if (kernel)
```
Comment (vcaputo) [inline grub-core/loader/i386/efi/linux.c]: unnecessary check, grub_free() noops if ptr is null

File: grub-core/loader/i386/linux.c [hunk 9]
```diff
@@ -1031,6 +1042,9 @@ grub_cmd_linux (grub_command_t cmd __attribute__ ((unused)),
 
  fail:
 
+  if (kernel)
```
Comment (vcaputo) [inline grub-core/loader/i386/linux.c]: unnecessary check

File: grub-core/loader/i386/pc/linux.c [hunk 10]
```diff
@@ -368,6 +377,9 @@ grub_cmd_linux (grub_command_t cmd __attribute__ ((unused)),
 
  fail:
 
+  if (kernel)
```
Comment (vcaputo) [inline grub-core/loader/i386/pc/linux.c]: unnecessary check...

File: grub-core/script/execute.c [hunk 11]
```diff
@@ -943,6 +945,25 @@ grub_script_execute_cmdline (struct grub_script_cmd *cmd)
...
+
+  cmdstring = grub_malloc (cmdlen);
+  if (!cmdstring)
+  {
+	  return grub_error (GRUB_ERR_OUT_OF_MEMORY,
+			     N_("cannot allocate command buffer"));
+  }
+
+  for (i = 0; i < argv.argc; i++) {
+	  offset += grub_snprintf(cmdstring + offset, cmdlen - offset, "%s ",
```
Comment (vcaputo) [inline grub-core/script/execute.c]: I don't know what's going on in grub, either always have a space after the function name or don't... it looks like the rest of this file has a space, so this and the call a few lines below should have spaces I guess.

Some of the listings seemed predominantly space-less, maybe those need to be changed to have spaces, *shrug*

**Comment by vcaputo** [comment]
made a fast pass, just some nits, will take a closer look next time around.

File: grub-core/boot/i386/pc/diskboot.S (hunk 1 above)
Comment (mjg59) [inline grub-core/boot/i386/pc/diskboot.S]: This was dragged in from the old Trusted Grub work, so I didn't alter it.

File: grub-core/kern/efi/tpm.c (hunk 3 above)
Comment (mjg59) [inline grub-core/kern/efi/tpm.c]: Yup

File: grub-core/kern/efi/tpm.c (hunk 2 above)
Comment (mjg59) [inline grub-core/kern/efi/tpm.c]: Sure

File: grub-core/kern/i386/pc/tpm.c (hunk 5 above)
Comment (mjg59) [inline grub-core/kern/i386/pc/tpm.c]: Fixed

File: grub-core/loader/i386/efi/linux.c (hunk 8 above)
Comment (mjg59) [inline grub-core/loader/i386/efi/linux.c]: Fixed

**Pull request closed without merging**